classes/UpgradeTools/CoreUpgrader/CoreUpgrader.php

```diff
@@ -583,105 +577,105 @@ protected function generateHtaccess(): void
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/cache/Cache.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/cache/Cache.php';
         }
         if (!class_exists('Cache', false) && class_exists('CacheCore')) {
             eval('abstract class Cache extends CacheCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/PrestaShopCollection.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/PrestaShopCollection.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/PrestaShopCollection.php';
         }
         if (!class_exists('PrestaShopCollection', false) && class_exists('PrestaShopCollectionCore')) {
             eval('class PrestaShopCollection extends PrestaShopCollectionCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/shop/ShopUrl.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/shop/ShopUrl.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/shop/ShopUrl.php';
         }
         if (!class_exists('ShopUrl', false) && class_exists('ShopUrlCore')) {
             eval('class ShopUrl extends ShopUrlCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/shop/Shop.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/shop/Shop.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/shop/Shop.php';
         }
         if (!class_exists('Shop', false) && class_exists('ShopCore')) {
             eval('class Shop extends ShopCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/Translate.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/Translate.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/Translate.php';
         }
         if (!class_exists('Translate', false) && class_exists('TranslateCore')) {
             eval('class Translate extends TranslateCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/module/Module.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/module/Module.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/module/Module.php';
         }
         if (!class_exists('Module', false) && class_exists('ModuleCore')) {
             eval('class Module extends ModuleCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/Validate.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/Validate.php')) {
             require_once _PS_ROOT_DIR_ . '/classes/Validate.php';
         }
         if (!class_exists('Validate', false) && class_exists('ValidateCore')) {
             eval('class Validate extends ValidateCore{}');
         }
 
-        if (file_exists(_PS_ROOT_DIR_ . '/classes/Language.php')) {
+        if ($this->container->getFileSystem()->exists(_PS_ROOT_DIR_ . '/classes/Language.php')) {
```

Comment:
It would have been better to store the instance in a $filesystem variable to avoid the overhead of calling the methods again and again